Provide `MockSelectionConstraints` type and generator.

jonathanknowles opened the pull request:
## Issue Number

ADP-1118
ADP-1136

## Summary

This PR:
- [x] Introduces the `MockSelectionConstraints` type with an associated generator and shrinker.
    The goal of this type is to:
    - generate arbitrary selection constraints to test `performSelection`;
    - provide good shrinking of these constraints.
- [x] Uses the `MockSelectionConstraints` type to simplify the `prop_performSelection` family of properties.
- [x] Moves `computeSelectionLimit` out of the `SelectionParams` record and into the `SelectionConstraints` record.
    This location is more appropriate, since:
    - `SelectionConstraints` is designed to collect constraints that are _common to all selections in the current era_, whereas `SelectionParams` is designed to collect parameters _specific to a given selection_.
    - `SelectionConstraints` is now a subset of `SelectionConstraints` in the parent module.

Additional work (to be used by a future PR):
- [x] Adds the `Coin.difference` function with property tests.
- [x] Adds the `TokenMap.size` function with property tests.

jonathanknowles committed:
Unwrap `assessTokenBundleSize` function within `SelectionConstraints`.

jonathanknowles committed:
Slightly re-arrange selection constraint sections in `BalanceSpec`.

jonathanknowles committed:
Remove layer of indirection when creating mock token bundle size assessors.

jonathanknowles committed:
Add generators and shrinkers for `MockAssessTokenBundleSize`.

jonathanknowles committed:
Clean up types, generators, and shrinkers for minimum cost functions.

jonathanknowles committed:
Clean up types, generators, and shrinkers for minimum ada quantity functions.

jonathanknowles committed:
Add `MockSelectionConstraints` type with generator and shrinker.

jonathanknowles committed:
Use `MockSelectionConstraints` to simplify `prop_performSelection`.

jonathanknowles committed:
Add `MockComputeSelectionLimit` type with generator and shrinker.

jonathanknowles committed:
Use `genMockComputeSelectionLimit` to simplify `genSelectionParams`.

jonathanknowles committed:
Add function `liftShrink4` to `Test.QuickCheck.Extra`.

jonathanknowles committed:
Move selection limit from `SelectionParams` to `SelectionConstraints`.

jonathanknowles committed:
Create dedicated `prop_performSelection_huge` property function.

jonathanknowles committed:
Remove outdated section in comment for `performSelection`.

jonathanknowles committed:
Add function `Coin.difference` to perform safe subtraction.

This function is equivalent to `Coin.subtract` but returns a default
value of `Coin 0` if the result would be negative.

This function is analogous to the following difference functions:

    - Map          .difference
    - Set          .difference
    - TokenBundle  .difference
    - TokenMap     .difference
    - TokenQuantity.difference
    - UTxO         .difference
    - UTxOIndex    .difference

jonathanknowles committed:
Remove `utxoRemaining` field from `SelectionResult`.

This field isn't really used anywhere except the test suite.

And it's arguably not part of what it means to be a "selection".

As for testing the contents of a `SelectionResult`, the same information
can already be derived from the `inputsSelected` field, which must be a
subset of `utxoAvailable`.

jonathanknowles committed:
Test that `assetsTo{Mint,Burn}` is preserved in the selection result.

jonathanknowles committed:
Use function `Coin.difference` to replace `fromMaybe (Coin 0) . subtract`.

jonathanknowles committed:
Add function `TokenMap.size` to compute the number of assets in a map.

jonathanknowles committed:
Use function `TokenMap.size` to replace `Set.size . TokenMap.getAssets`.

sevanspowell commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/BalanceSpec.hs:
```diff
@@ -1939,60 +1999,136 @@ mkMinCoinValueFor = \case
-    NoBundleSizeLimit ->
+genMockAssessTokenBundleSize :: Gen MockAssessTokenBundleSize
+genMockAssessTokenBundleSize =
+    -- TODO:
```
Is there a reason why we don't?

jonathanknowles replied:
> Is there a reason why we don't?

If I recall correctly, at the time we wrote `prop_performSelection` we believed it would complicate the property test too much to impose a token bundle size limit, so we always generated criteria without them.

I'm actually planning to improve this generator in the next PR: https://github.com/input-output-hk/cardano-wallet/pull/2909

jonathanknowles committed:
Add property tests for function `Coin.difference`.

We also add other property tests to make up for missing coverage of
arithmetic operations on `Coin`.

jonathanknowles committed:
Add property tests for function `TokenMap.size`.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Balance.hs:
```diff
@@ -411,10 +411,6 @@ data SelectionResult change = SelectionResult
     , changeGenerated
         :: ![change]
         -- ^ A list of generated change outputs.
-    , utxoRemaining
```
:clap:
